New time period module

Max-checkmk opened the pull request:
<!--- Please provide a brief summary of your changes as a title in the textbox above -->

<!---
Please use the devel branch as the merge target (see dropdown above)!
We use that branch as a staging area, to make sure the main branch
stays as stable as possible, in case people are using it to install the collection directly.
 -->

## Pull request type
<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

There is no module for time periods

## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->

- Added a module for creating, updating and deleting time periods

## Other information
<!-- Any other information that is important to this PR, e.g screenshots of how the component looks before and after the change. -->

lgetwan reviewed (changes requested):
The documentation could be a bit more detailed. Probably, you could use the tasks from your integration test.

I tested the module with some use cases:
- When creating a timeperiod that already exists, a failure is shown (should be "OK").
- When updating a timeperiod without any actual change, the result is "changed" (should be "OK").
- When adding an exclude time range, this isn't visible in the UI.
- When deleting a timeperiod that doesn't exist, a failure is shown (should be "OK").

Here's the playbook i used:
```---

# Make sure to keep the tests in sync!
# Rather use the tests than this playbook directly.

- name: "Test timeperiods"
  hosts: all
  gather_facts: 'no'
  vars_files:
    - ./vars/config.yml

  tasks:
  - name: create timeperiod
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "lunch"
      alias: "Mahlzeit"
      active_time_ranges: [
              { 
                  "day": "all",
                  "time_ranges": [
                      {
                          "start": "12:00",
                          "end": "13:00"
                      }
                  ]
              }
          ]
      state: present
    delegate_to: localhost
    run_once: 'yes'

  # fails. should be "OK"
  - name: create timeperiod again
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "lunch"
      alias: "Mahlzeit"
      active_time_ranges: [
              { 
                  "day": "all",
                  "time_ranges": [
                      {
                          "start": "12:00",
                          "end": "13:00"
                      }
                  ]
              }
          ]
      state: present
    delegate_to: localhost
    run_once: 'yes'

  - name: create timeperiod that uses the first one
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "working_hours"
      alias: "Ballern"
      active_time_ranges: [
              { 
                  "day": "all",
                  "time_ranges": [
                      {
                          "start": "8:00",
                          "end": "17:00"
                      }
                  ]
              },
          ]
      exclude: ["Mahlzeit"]
      state: present
    delegate_to: localhost
    run_once: 'yes'

  - name: update timeperiod
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "working_hours"
      active_time_ranges: [
              { 
                  "day": "all",
                  "time_ranges": [
                      {
                          "start": "9:00",
                          "end": "18:00"
                      }
                  ]
              },
          ]
      state: present
    delegate_to: localhost
    run_once: 'yes'

  # fails. should be "OK"
  - name: update timeperiod. again.
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "working_hours"
      active_time_ranges: [
              { 
                  "day": "all",
                  "time_ranges": [
                      {
                          "start": "9:00",
                          "end": "18:00"
                      }
                  ]
              },
          ]
      state: present
    delegate_to: localhost
    run_once: 'yes'

  - name: remove used timeperiod
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "lunch"
      state: absent
    delegate_to: localhost
    run_once: 'yes'
    ignore_errors: true

  - name: remove timeperiod
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "working_hours"
      state: absent
    delegate_to: localhost
    run_once: 'yes'

  # fails. should be "OK"
  - name: remove timeperiod again
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "working_hours"
      state: absent
    delegate_to: localhost
    run_once: 'yes'

  - name: remove used timeperiod
    checkmk.general.timeperiod:
      server_url: "{{ server_url }}"
      site: "{{ site }}"
      automation_user: "{{ automation_user }}"
      automation_secret: "{{ automation_secret }}"
      name: "lunch"
      state: absent
    delegate_to: localhost
    run_once: 'yes'
```

Max-checkmk commented:
> The documentation could be a bit more detailed. Probably, you could use the tasks from your integration test.
> 
> I tested the module with some use cases:
> 
>     * When creating a timeperiod that already exists, a failure is shown (should be "OK").
> 
>     * When updating a timeperiod without any actual change, the result is "changed" (should be "OK").
> 
>     * When adding an exclude time range, this isn't visible in the UI.
> 
>     * When deleting a timeperiod that doesn't exist, a failure is shown (should be "OK").
> 
> 
> ```

Hi Lars,

thanks for the feedback. I had a dumb mistake in it, which is now corrected.
And i forgot to finish the examples, which is also done now.

Please try it again.
What i didn't change is the error which is thrown, if someones tries to delete something what doesn't exist.
Not sure about this...

Nice weekend! ;)

msekania commented:
> What i didn't change is the error which is thrown, if someones tries to delete something what doesn't exist.
> Not sure about this...

@Max-checkmk,

`state: absent` means that it's not there, that means that someone does not try to delete something but rather to ensure that it's not there, hence no error.

lgetwan commented on plugins/modules/timeperiod.py:
```diff
@@ -0,0 +1,444 @@
+                )
+                module.fail_json(**result_as_dict(result))
+
+            timeperiodupdate = TimeperiodUpdateAPI(module)
+            timeperiodupdate.headers["If-Match"] = result.etag
```
An Update should only be done in the case the mentioned parameters are different in the already existing Timeperiod. 
Otherwise running the same playbook again and again will always do a change in the environment. You have to avoid this.

Max-checkmk replied:
Nice idea, but this will need a lot of time to accomplish.

So as example we want to update a time period for all weekdays with these time ranges:
`active_time_ranges: '[{"day": "all", "time_ranges": [{"start": "08:00:00", "end": "17:00:00"}]}]'`

And we try to compare this with the existing time ranges of a time period, we get this:
`[{'day': 'monday', 'time_ranges': [{'start': '08:00', 'end': '17:00'}]}, {'day': 'tuesday', 'time_ranges': [{'start': '08:00', 'end': '17:00'}]}, {'day': 'wednesday', 'time_ranges': [{'start': '08:00', 'end': '17:00'}]}, {'day': 'thursday', 'time_ranges': [{'start': '08:00', 'end': '17:00'}]}, {'day': 'friday', 'time_ranges': [{'start': '08:00', 'end': '17:00'}]}, {'day': 'saturday', 'time_ranges': [{'start': '08:00', 'end': '17:00'}]}, {'day': 'sunday', 'time_ranges': [{'start': '08:00', 'end': '17:00'}]}]`

So we would need to adjust
- the given time ranges to every single day of the week
- remove the seconds from the input, as internally there are no seconds (in contrast to the API Docs example...)

We redirect the raw data to the API so at first you would have to convert the data to something you can work with in python.

lgetwan replied:
I see, that's too complex.

But what do you think of this as an intermediate solution:
If the data provided by the task is completely equal with the current data from the API, we don't change anything. But if the task e.g. has seconds in the data, and the API has not, we do post a change to the API.

Many users are running their playbooks daily or hourly, and that would help them a lot, already.

Max-checkmk replied:
Please have a look at my last commit. Now there is a rudimentary comparison between existing and new values.
Based on that we could add the other checks (full week and seconds) sometime.

lgetwan commented:
Nice weekent y'all!

Max-checkmk commented:
> `state: absent` means that it's not there, that means that someone does not try to delete something but rather to ensure that it's not there, hence no error, rather no change. And changed if it was there and deleted.

You're right. I just had a look at some other modules and as far i see at the moment we're not really consistent for all modules here. ;)
I will do the change for this module with the next commit!

lgetwan reviewed (approved):
Works as described. :-)

robin-checkmk commented:
@Max-checkmk please check the failing sanity check and fix it. I think we had this issue with the host or folder module in the past. Maybe @lgetwan can help here.

Max-checkmk commented:
> @Max-checkmk please check the failing sanity check and fix it. I think we had this issue with the host or folder module in the past. Maybe @lgetwan can help here.

Well, the problem seems to be inside the folder-module and not in this module. As far as i can read the logs from the test...

robin-checkmk commented:
True. Then we will ignore the failing test here and fix it in `devel` directly.
